### PR TITLE
Disentangle WalkerControl, BranchEngine and EstimatorManager in batched drivers

### PR DESCRIPTION
**Pull request by ye-luo**
## Proposed changes
Review after #2866. Expect CI test run failure until #2867 gets a fix.
WalkerControl is moved out of BranchEngine.
BranchEngine no more keep EstimatorManagerNew pointer.
WalkerControl, BranchEngine is moved from QMCDriverNew base class to DMCBatch.

Also add `const ProjectData& project_info_;` to QMCDriverNew for handling section counter and file root.

## What type(s) of changes does this code introduce?
- Code style update (formatting, renaming)
- Refactoring (no functional changes, no api changes)

### Does this introduce a breaking change?
- No

## What systems has this change been tested on?
epyc-server

## Checklist
- Yes. This PR is up to date with current the current state of 'develop'

### CODE REVIEW

**Review by PDoakORNL — CHANGES_REQUESTED** [review]
Some nice disentanglement here.  I wouldn't say SFNBranch is easily understood yet but this is a big step.

I'd like to see some naming and removal of state from the driver factory but that's optional.
The batched determinant magic number replacement/documentation is important.

File: src/QMCDrivers/DMC/DMCBatched.cpp [hunk 1]
```diff
@@ -478,11 +450,40 @@ void DMCBatched::process(xmlNodePtr node)
...
+    ReportEngine PRE("DMC", "resetUpdateEngines");
+    Timer init_timer;
+    // Here DMC loads "Ensemble of cloned MCWalkerConfigurations"
+    // I'd like to do away with this method in DMCBatched.
+
+    app_log() << "  Creating the branching engine and walker controler" << std::endl;
+    branch_engine_ = std::make_unique<SFNBranch>(qmcdriver_input_.get_tau(), population_.get_num_global_walkers());
+    branch_engine_->put(node);
+
+    WalkerController = std::make_unique<WalkerControl>(myComm, Random, dmcdriver_input_.get_reconfiguration());
```
[the remark this answers is not in the document]
Comment (ye-luo) [reply src/QMCDrivers/DMC/DMCBatched.cpp:464]: Agreed. Will fix. Momentum from the old code.

File: src/QMCDrivers/DMC/DMCBatched.h [hunk 2]
```diff
@@ -111,18 +120,17 @@ class DMCBatched : public QMCDriverNew
...
   IndexType branch_interval_;
-  void resetUpdateEngines();
-  /// Copy Constructor (disabled)
-  DMCBatched(const DMCBatched&) = delete;
-  /// Copy operator (disabled).
-  DMCBatched& operator=(const DMCBatched&) = delete;
+  ///branch engine
+  std::unique_ptr<SFNBranch> branch_engine_;
+  ///walker controller for load-balance
+  std::unique_ptr<WalkerControl> WalkerController;
```
[the remark this answers is not in the document]
Comment (ye-luo) [reply src/QMCDrivers/DMC/DMCBatched.h:133]: Will fix.

File: src/QMCDrivers/DMC/WalkerControl.cpp [hunk 3]
```diff
@@ -57,15 +58,13 @@ WalkerControl::WalkerControl(Communicate* c, RandomGenerator_t& rng, bool use_fi
       n_min_(1),
       n_max_(10),
       max_copy_(2),
-      target_sigma_(10),
       dmcStream(0),
+      rank_num_(c->rank()),
```
[the remark this answers is not in the document]
Comment (ye-luo) [reply src/QMCDrivers/DMC/WalkerControl.cpp:62]: Writing tons of myComm->rank() also annoyed me. So I'd like to keep it as it is. I have used const in both num_ranks_ and rank_num_ to ensure them set only by the constructor.

File: src/QMCDrivers/QMCDriverFactory.cpp [hunk 4]
```diff
@@ -242,8 +244,9 @@ std::unique_ptr<QMCDriverInterface> QMCDriverFactory::createQMCDriver(xmlNodePtr
   }
   else if (das.new_run_type == QMCRunType::VMC_BATCH)
   {
-    VMCFactoryNew fac(cur, das.what_to_do[UPDATE_MODE], qmc_common.qmc_counter);
-    new_driver.reset(fac.create(MCPopulation(comm->size(), comm->rank(), qmc_system, &qmc_system, primaryPsi, primaryH),
+    VMCFactoryNew fac(cur, das.what_to_do[UPDATE_MODE]);
+    new_driver.reset(fac.create(project_info_,
```
[the remark this answers is not in the document]
Comment (ye-luo) [reply src/QMCDrivers/QMCDriverFactory.cpp:248]: A driver factory is responsible for all the driver creation. das seems to be used only for reading a specific driver attributes.  I'm not sure it is necessary to move pass it via das.

File: src/QMCDrivers/QMCDriverFactory.h [hunk 5]
```diff
@@ -64,6 +67,10 @@ class QMCDriverFactory
                                                       WaveFunctionPool& wave_function_pool,
                                                       HamiltonianPool& hamiltonian_pool,
                                                       Communicate* comm) const;
+
+private:
+  /// project info for accessing global fileroot and series id
+  const ProjectData& project_info_;
```
[the remark this answers is not in the document]
Comment (ye-luo) [reply src/QMCDrivers/QMCDriverFactory.h:73]: I don't really see the benefit of passing project_data_ in another way so far.
Renaming yes, I will change it throughout the code.

File: src/QMCWaveFunctions/Fermion/DiracDeterminantBatched.cpp [hunk 6]
```diff
@@ -345,6 +329,46 @@ typename DiracDeterminantBatched<DET_ENGINE_TYPE>::LogValueType DiracDeterminant
...
+      RefVector<DET_ENGINE_TYPE> engine_list;
+      engine_list.reserve(WFC_list.size());
+
+      for (int iw = 0; iw < WFC_list.size(); iw++)
+      {
+        auto& det = static_cast<DiracDeterminantBatched<DET_ENGINE_TYPE>&>(WFC_list[iw].get());
+        engine_list.push_back(det.det_engine_);
+        auto& my_psiM_vgl  = det.psiM_vgl;
+        auto* psiM_vgl_ptr = my_psiM_vgl.data();
+        PRAGMA_OFFLOAD("omp target update from(psiM_vgl_ptr[my_psiM_vgl.capacity():my_psiM_vgl.capacity()*4]) nowait")
```
[the remark this answers is not in the document]
Comment (ye-luo) [reply src/QMCWaveFunctions/Fermion/DiracDeterminantBatched.cpp:355]: I will add a comment.

**Review by ye-luo — COMMENTED** [review]
@PDoakORNL thank you for the careful review. I'm currently busy with some other debugging. Consider the issues are all minor can we merge this to unblock #2875? Reviewing my update on a separate PR also easier than reviewing this PR again the updates in place.

**Review by PDoakORNL — APPROVED** [review]

**Pull request merged**